File: pkg/dev_compiler/test/codegen_expected/language/class_test.js
dart_library.library('language/class_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__class_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const class_test = Object.create(null);
  let intToint = () => (intToint = dart.constFn(dart.definiteFunctionType(core.int, [core.int])))();
  let VoidToint = () => (VoidToint = dart.constFn(dart.definiteFunctionType(core.int, [])))();
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  class_test.ClassTest = class ClassTest extends core.Object {
    new() {
    }
    static testMain() {
      let test = new class_test.ClassTest();
      test.testSuperCalls();
      test.testVirtualCalls();
      test.testStaticCalls();
      test.testInheritedField();
      test.testMemberRefInClosure();
      test.testFactory();
      test.testNamedConstructors();
      test.testDefaultImplementation();
      test.testFunctionParameter(dart.fn(a => a, intToint()));
    }
    testFunctionParameter(func) {
      expect$.Expect.equals(1, func(1));
    }
    testSuperCalls() {
      let sub = new class_test.Sub();
      expect$.Expect.equals(43, sub.methodX());
      expect$.Expect.equals(84, sub.methodK());
    }
    testVirtualCalls() {
      let sub = new class_test.Sub();
      expect$.Expect.equals(41, sub.method2());
      expect$.Expect.equals(41, sub.method3());
    }
    testStaticCalls() {
      let sub = new class_test.Sub();
      expect$.Expect.equals(-42, class_test.Sub.method4());
      expect$.Expect.equals(-41, sub.method5());
    }
    testInheritedField() {
      let sub = new class_test.Sub();
      expect$.Expect.equals(42, sub.method6());
    }
    testMemberRefInClosure() {
      let sub = new class_test.Sub();
      expect$.Expect.equals(1, sub.closureRef());
      expect$.Expect.equals(2, sub.closureRef());
      sub = new class_test.Sub();
      expect$.Expect.equals(1, sub.closureRef());
      expect$.Expect.equals(2, sub.closureRef());
    }
    testFactory() {
      let sup = class_test.Sup.named();
      expect$.Expect.equals(43, sup.methodX());
      expect$.Expect.equals(84, sup.methodK());
    }
    testNamedConstructors() {
      let sup = new class_test.Sup.fromInt(4);
      expect$.Expect.equals(4, sup.methodX());
      expect$.Expect.equals(0, sup.methodK());
    }
    testDefaultImplementation() {
      let x = class_test.Inter.new(4);
      expect$.Expect.equals(4, x.methodX());
      expect$.Expect.equals(8, x.methodK());
      x = class_test.Inter.fromInt(4);
      expect$.Expect.equals(4, x.methodX());
      expect$.Expect.equals(0, x.methodK());
      x = class_test.Inter.named();
      expect$.Expect.equals(43, x.methodX());
      expect$.Expect.equals(84, x.methodK());
      x = class_test.Inter.factory();
      expect$.Expect.equals(43, x.methodX());
      expect$.Expect.equals(84, x.methodK());
    }
  };
  dart.setSignature(class_test.ClassTest, {
    constructors: () => ({new: dart.definiteFunctionType(class_test.ClassTest, [])}),
    methods: () => ({
      testFunctionParameter: dart.definiteFunctionType(dart.dynamic, [dart.functionType(core.int, [core.int])]),
      testSuperCalls: dart.definiteFunctionType(dart.dynamic, []),
      testVirtualCalls: dart.definiteFunctionType(dart.dynamic, []),
      testStaticCalls: dart.definiteFunctionType(dart.dynamic, []),
      testInheritedField: dart.definiteFunctionType(dart.dynamic, []),
      testMemberRefInClosure: dart.definiteFunctionType(dart.dynamic, []),
      testFactory: dart.definiteFunctionType(dart.dynamic, []),
      testNamedConstructors: dart.definiteFunctionType(dart.dynamic, []),
      testDefaultImplementation: dart.definiteFunctionType(dart.dynamic, [])
    }),
    statics: () => ({testMain: dart.definiteFunctionType(dart.dynamic, [])}),
    names: ['testMain']
  });
  class_test.Inter = class Inter extends core.Object {
    static named() {
      return class_test.Sup.named();
    }
    static fromInt(x) {
      return new class_test.Sup.fromInt(x);
    }
    static new(x) {
      return new class_test.Sup(x);
    }
    static factory() {
      return class_test.Sup.factory();
    }
  };
  dart.setSignature(class_test.Inter, {
    constructors: () => ({
      named: dart.definiteFunctionType(class_test.Inter, []),
      fromInt: dart.definiteFunctionType(class_test.Inter, [core.int]),
      new: dart.definiteFunctionType(class_test.Inter, [core.int]),
      factory: dart.definiteFunctionType(class_test.Inter, [])
    })
  });
  class_test.Sup = class Sup extends core.Object {
    static named() {
      return new class_test.Sub();
    }
    static factory() {
      return new class_test.Sub();
    }
    fromInt(x) {
      this.x_ = null;
      this.k_ = null;
      this.x_ = x;
      this.k_ = 0;
    }
    methodX() {
      return this.x_;
    }
    methodK() {
      return this.k_;
    }
    new(x) {
      this.x_ = x;
      this.k_ = null;
      this.k_ = dart.notNull(x) * 2;
    }
    method2() {
      return dart.notNull(this.x_) - 1;
    }
  };
  dart.defineNamedConstructor(class_test.Sup, 'fromInt');
  class_test.Sup[dart.implements] = () => [class_test.Inter];
  dart.setSignature(class_test.Sup, {
    constructors: () => ({
      named: dart.definiteFunctionType(class_test.Sup, []),
      factory: dart.definiteFunctionType(class_test.Sup, []),
      fromInt: dart.definiteFunctionType(class_test.Sup, [core.int]),
      new: dart.definiteFunctionType(class_test.Sup, [core.int])
    }),
    methods: () => ({
      methodX: dart.definiteFunctionType(core.int, []),
      methodK: dart.definiteFunctionType(core.int, []),
      method2: dart.definiteFunctionType(core.int, [])
    })
  });
  class_test.Sub = class Sub extends class_test.Sup {
    methodX() {
      return dart.notNull(super.methodX()) + 1;
    }
    method3() {
      return this.method2();
    }
    static method4() {
      return -42;
    }
    method5() {
      return dart.notNull(class_test.Sub.method4()) + 1;
    }
    method6() {
      return dart.notNull(this.x_) + dart.notNull(this.y_);
    }
    closureRef() {
      let f = dart.fn(() => {
        this.y_ = dart.notNull(this.y_) + 1;
        return this.y_;
      }, VoidToint());
      return f();
    }
    new() {
      this.y_ = null;
      super.new(42);
      this.y_ = 0;
    }
  };
  dart.setSignature(class_test.Sub, {
    constructors: () => ({new: dart.definiteFunctionType(class_test.Sub, [])}),
    methods: () => ({
      method3: dart.definiteFunctionType(core.int, []),
      method5: dart.definiteFunctionType(core.int, []),
      method6: dart.definiteFunctionType(core.int, []),
      closureRef: dart.definiteFunctionType(core.int, [])
    }),
    statics: () => ({method4: dart.definiteFunctionType(core.int, [])}),
    names: ['method4']
  });
  class_test.main = function() {
    class_test.ClassTest.testMain();
  };
  dart.fn(class_test.main, VoidTodynamic());
  // Exports:
  exports.class_test = class_test;
});

File: tests/language/class_test.dart
// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import "package:expect/expect.dart";

// Tests basic classes and methods.
class ClassTest {
  ClassTest() {}

  static testMain() {
    var test = new ClassTest();
    test.testSuperCalls();
    test.testVirtualCalls();
    test.testStaticCalls();
    test.testInheritedField();
    test.testMemberRefInClosure();
    test.testFactory();
    test.testNamedConstructors();
    test.testDefaultImplementation();
    test.testFunctionParameter((int a) {
      return a;
    });
  }

  testFunctionParameter(int func(int a)) {
    Expect.equals(1, func(1));
  }

  testSuperCalls() {
    var sub = new Sub();
    Expect.equals(43, sub.methodX());
    Expect.equals(84, sub.methodK());
  }

  testVirtualCalls() {
    var sub = new Sub();
    Expect.equals(41, sub.method2());
    Expect.equals(41, sub.method3());
  }

  testStaticCalls() {
    var sub = new Sub();
    Expect.equals(-42, Sub.method4());
    Expect.equals(-41, sub.method5());
  }

  testInheritedField() {
    var sub = new Sub();
    Expect.equals(42, sub.method6());
  }

  testMemberRefInClosure() {
    var sub = new Sub();
    Expect.equals(1, sub.closureRef());
    Expect.equals(2, sub.closureRef());
    // Make sure it is actually on the object, not the global 'this'.
    sub = new Sub();
    Expect.equals(1, sub.closureRef());
    Expect.equals(2, sub.closureRef());
  }

  testFactory() {
    var sup = new Sup.named();
    Expect.equals(43, sup.methodX());
    Expect.equals(84, sup.methodK());
  }

  testNamedConstructors() {
    var sup = new Sup.fromInt(4);
    Expect.equals(4, sup.methodX());
    Expect.equals(0, sup.methodK());
  }

  testDefaultImplementation() {
    var x = new Inter(4);
    Expect.equals(4, x.methodX());
    Expect.equals(8, x.methodK());

    x = new Inter.fromInt(4);
    Expect.equals(4, x.methodX());
    Expect.equals(0, x.methodK());

    x = new Inter.named();
    Expect.equals(43, x.methodX());
    Expect.equals(84, x.methodK());

    x = new Inter.factory();
    Expect.equals(43, x.methodX());
    Expect.equals(84, x.methodK());
  }
}

abstract class Inter {
  factory Inter.named() = Sup.named;
  factory Inter.fromInt(int x) = Sup.fromInt;
  factory Inter(int x) = Sup;
  factory Inter.factory() = Sup.factory;
  int methodX();
  int methodK();
  int x_;
}

class Sup implements Inter {
  int x_;
  int k_;

  factory Sup.named() {
    return new Sub();
  }

  factory Sup.factory() {
    return new Sub();
  }

  Sup.fromInt(int x) {
    x_ = x;
    k_ = 0;
  }

  int methodX() {
    return x_;
  }

  int methodK() {
    return k_;
  }

  Sup(int x) : this.x_ = x {
    k_ = x * 2;
  }

  int method2() {
    return x_ - 1;
  }
}

class Sub extends Sup {
  int y_;

  // Override
  int methodX() {
    return super.methodX() + 1;
  }

  int method3() {
    return method2();
  }

  static int method4() {
    return -42;
  }

  int method5() {
    return method4() + 1;
  }

  int method6() {
    return x_ + y_;
  }

  int closureRef() {
    var f = () {
      y_ += 1;
      return y_;
    };
    return f();
  }

  Sub() : super(42) {
    y_ = 0;
  }
}

main() {
  ClassTest.testMain();
}
